src/Main.elm

```diff
@@ -1,10 +1,829 @@
+    , currentFolderPath = []
+    , sortBy = Data.Audience.Authored
+    }
+
+
+tryLoadingModelToLoaded : LoadingModel -> Model
+tryLoadingModelToLoaded loadingM =
+    case ( loadingM.audiences, loadingM.audienceFolders ) of
+        ( Just loadedAudiences, Just loadedAudienceFolders ) ->
+            Loaded (initLoadedModel loadedAudiences loadedAudienceFolders)
+
+        ( _, _ ) ->
+            Loading loadingM
+
+
+
+-- normally, here I would return a tuple of type (Model, Cmd Msg), but since
+-- in this app there's no Cmd s to run, I decided to simplify everything
+-- and just return the Model
+
+
+update : Msg -> Model -> Model
+update msg model =
+    case ( msg, model ) of
+        ( GotAudiencesJSON json, Loading loadingModel ) ->
+            let
+                audiencesDecoder : Json.Decode.Decoder (List Data.Audience.Audience)
+                audiencesDecoder =
+                    Json.Decode.field "data" <|
+                        Json.Decode.list Data.Audience.audienceDecoder
+            in
+            case Json.Decode.decodeString audiencesDecoder json of
+                Err err ->
+                    JsonError err
+
+                Ok audiencesList ->
+                    tryLoadingModelToLoaded
+                        { loadingModel
+                            | audiences = Just (audiencesToDict audiencesList)
+                        }
+
+        ( GotAudienceFoldersJSON json, Loading loadingModel ) ->
+            let
+                audienceFoldersDecoder : Json.Decode.Decoder (List Data.AudienceFolder.AudienceFolder)
+                audienceFoldersDecoder =
+                    Json.Decode.field "data" <|
+                        Json.Decode.list Data.AudienceFolder.audienceFolderDecoder
+            in
+            case Json.Decode.decodeString audienceFoldersDecoder json of
+                Err err ->
+                    JsonError err
+
+                Ok audienceFoldersList ->
+                    tryLoadingModelToLoaded
+                        { loadingModel
+                            | audienceFolders = Just (audienceFoldersToDict audienceFoldersList)
+                        }
+
+        ( FolderClicked sel, Loaded loadedModel ) ->
+            Loaded
+                { loadedModel
+                    | currentFolderPath = sel :: loadedModel.currentFolderPath
+                }
+
+        ( GoBack, Loaded loadedModel ) ->
+            Loaded
+                { loadedModel
+                    | currentFolderPath = List.drop 1 loadedModel.currentFolderPath
+                }
+
+        ( SortBy audienceType, Loaded loadedModel ) ->
+            Loaded
+                { loadedModel
+                    | sortBy = audienceType
+                    , currentFolderPath = []
+                }
+
+        ( GotTime time, Loading loadingModel ) ->
+            Loading
+                { loadingModel
+                    | time = Just time
+                }
+
+        ( _, _ ) ->
```

Comment:
This is always danger for future maintenance, you can easily forget cover some new combination.